auth.php

```diff
@@ -525,8 +533,15 @@ public function saml_login() {
         } else if (!is_null($saml2auth->defaultidp)) {
             $SESSION->saml2idp = md5($saml2auth->defaultidp->entityid);
         } else if ($saml2auth->multiidp) {
-            $idpurl = new moodle_url('/auth/saml2/selectidp.php');
-            redirect($idpurl);
+            // At this stage there is no alias, get-param or default IdP configured.
+            // On a multi-idp system, now check for any whitelisted IP address redirection.
+            $entitiyid = $this->check_whitelisted_ip_redirect();
+            if ($entitiyid !== null) {
+                $SESSION->saml2idp = $entitiyid;
+            } else {
+                $idpurl = new moodle_url('/auth/saml2/selectidp.php');
+                redirect($idpurl);
```

Comment:
shouldn't this be a string return not a redirect? + data provider

Reply:
This is the saml_login() function, this has no check for the return type in loginpage_hook(). Want me to return here and add more login in the hook?

Reply:
ah yup I got mixed up, leave this as is